lnd_test.go

```diff
@@ -574,88 +655,223 @@ func testMultiHopPayments(net *networkHarness, t *testing.T) {
-			}
-			if err := lightningNetwork.TearDownAll(); err != nil {
-				fmt.Println("unable to tear lnd harnesses: ", err)
-			}
-			t.Fatalf("test %v panicked: %s", currentTest, debug.Stack())
-		}
-	}()
+	ct := NewCT(t)
 
 	// First create the network harness to gain access to its
 	// 'OnTxAccepted' call back.
-	lightningNetwork, err = newNetworkHarness()
+	lndHarness, err := newNetworkHarness()
 	if err != nil {
-		t.Fatalf("unable to create lightning network harness: %v", err)
+		ct.Fatalf("unable to create lightning network harness: %v", err)
 	}
-	defer lightningNetwork.TearDownAll()
+	defer lndHarness.TearDownAll()
 
 	handlers := &btcrpcclient.NotificationHandlers{
-		OnTxAccepted: lightningNetwork.OnTxAccepted,
+		OnTxAccepted: lndHarness.OnTxAccepted,
 	}
 
 	// First create an instance of the btcd's rpctest.Harness. This will be
 	// used to fund the wallets of the nodes within the test network and to
 	// drive blockchain related events within the network.
-	btcdHarness, err = rpctest.New(harnessNetParams, handlers, nil)
+	btcdHarness, err := rpctest.New(harnessNetParams, handlers, nil)
 	if err != nil {
-		t.Fatalf("unable to create mining node: %v", err)
+		ct.Fatalf("unable to create mining node: %v", err)
 	}
 	defer btcdHarness.TearDown()
-	if err = btcdHarness.SetUp(true, 50); err != nil {
-		t.Fatalf("unable to set up mining node: %v", err)
+	if err := btcdHarness.SetUp(true, 50); err != nil {
+		ct.Fatalf("unable to set up mining node: %v", err)
 	}
 	if err := btcdHarness.Node.NotifyNewTransactions(false); err != nil {
-		t.Fatalf("unable to request transaction notifications: %v", err)
+		ct.Fatalf("unable to request transaction notifications: %v", err)
 	}
 
 	// With the btcd harness created, we can now complete the
 	// initialization of the network. args - list of lnd arguments,
 	// example: "--debuglevel=debug"
 	// TODO(roasbeef): create master balanced channel with all the monies?
-	if err := lightningNetwork.InitializeSeedNodes(btcdHarness, nil); err != nil {
-		t.Fatalf("unable to initialize seed nodes: %v", err)
+	if err := lndHarness.InitializeSeedNodes(btcdHarness, nil); err != nil {
+		ct.Fatalf("unable to initialize seed nodes: %v", err)
 	}
-	if err = lightningNetwork.SetUp(); err != nil {
-		t.Fatalf("unable to set up test lightning network: %v", err)
+	if err = lndHarness.SetUp(); err != nil {
+		ct.Fatalf("unable to set up test lightning network: %v", err)
 	}
 
-	t.Logf("Running %v integration tests", len(lndTestCases))
-	for testName, lnTest := range lndTestCases {
-		t.Logf("Executing test %v", testName)
+	ct.Logf("Running %v integration tests", len(testCases))
 
-		currentTest = testName
-		lnTest(lightningNetwork, t)
+out:
+	for name, test := range testCases {
+		errChan := ct.RunTest(lndHarness, test)
+
+		select {
+		// Receive both types of err - panic and fatal from
+		// one channel and raise the fatal in main "thread".
+		case err := <- errChan:
```

Comment:
A two-value read can be done here (`err, ok`) in order to check if the channel was closed, or an error was propagated up. So when `!ok` you know the test succeeded, and can move forwards.